src/v/serde/avro/parser.cc

```diff
@@ -0,0 +1,180 @@
+            auto record = std::make_unique<parsed::record>();
+
+            for (auto i : boost::irange<size_t>(node->leaves())) {
+                auto field_node = node->leafAt(i);
+                record->fields.push_back(
+                  co_await parse_node(field_node, state));
+            }
+            state.level--;
+            co_return std::make_unique<parsed::message>(std::move(record));
+        }
+        case ::avro::AVRO_ARRAY: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::list>();
+
+            if (element_count > 0) {
+                ret->elements.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    ret->elements.push_back(
+                      co_await parse_node(node->leafAt(0), state));
+                }
+                // avro always writes 0 as the array footer, first element count
+                // field is skipped if array is empty
+                _parser.read_varlong();
+            }
+            state.level--;
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_MAP: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::map>();
+            if (element_count > 0) {
+                ret->entries.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    auto key = parse_string();
+                    ret->entries.emplace_back(
+                      std::move(key),
+                      co_await parse_node(node->leafAt(1), state));
+                }
+                // avro always writes 0 as the map footer, first element count
+                // field is skipped if the map is empty
+                _parser.read_varlong();
+            }
+            state.level--;
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_UNION: {
+            auto idx = parse_length();
+            auto union_branch = node->leafAt(idx);
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              std::make_unique<parsed::avro_union>(parsed::avro_union{
+                .branch = static_cast<size_t>(idx),
+                .message = co_await parse_node(union_branch, state)}));
+        }
+        case ::avro::AVRO_FIXED: {
+            // fixed type is special as it does not have the size included in
+            // the binary representation
+            state.level--;
+            co_return std::make_unique<parsed::message>(
+              _parser.share(node->fixedSize()));
+        }
+        case ::avro::AVRO_SYMBOLIC:
+            // this is a special case used when dealing with nested types, we
+            // need to resolve the original type and parse it
+            co_return co_await parse_node(::avro::resolveSymbol(node), state);
```

Comment:
Is it intentional we don't `state.level--` here?

If so, it warrants a comment, as this is the only branch that doesn't do this. If that's a bug - can we use `absl::Cleanup` or `ss::defer` to decrement the level instead of having to remember to do this in every branch?

Reply:
this is where a level gets deeper actually, as this indicates the recursive type so i haven't decrement it here on purpose, i will use defer

Reply:
actually the level should be decremented after the `parse_node` returned